documents/read_acls.go

```diff
@@ -0,0 +1,336 @@
+func (cd *CoreDocument) CreateNFTProofs(
+	docType string,
+	account identity.CentID,
+	registry common.Address,
+	tokenID []byte,
+	nftUniqueProof, readAccessProof bool) (proofs []*proofspb.Proof, err error) {
+
+	if len(cd.Document.DataRoot) != idSize {
+		return nil, errors.New("data root is invalid")
+	}
+
+	var pfKeys []string
+	if nftUniqueProof {
+		pk, err := getNFTUniqueProofKey(cd.Document.Nfts, registry)
+		if err != nil {
+			return nil, err
+		}
+
+		pfKeys = append(pfKeys, pk)
+	}
+
+	if readAccessProof {
+		pks, err := getReadAccessProofKeys(cd.Document, registry, tokenID)
+		if err != nil {
+			return nil, err
+		}
+
+		pfKeys = append(pfKeys, pks...)
+	}
+
+	signingRootProofHashes, err := cd.getSigningRootProofHashes()
+	if err != nil {
+		return nil, errors.New("failed to generate signing root proofs: %v", err)
+	}
+
+	cdTree, err := cd.documentTree(docType)
+	if err != nil {
+		return nil, errors.New("failed to generate core Document tree: %v", err)
+	}
+
+	proofs, missedProofs := generateProofs(cdTree, pfKeys, append([][]byte{cd.Document.DataRoot}, signingRootProofHashes...))
+	if len(missedProofs) != 0 {
+		return nil, errors.New("failed to create proofs for fields %v", missedProofs)
+	}
+
+	return proofs, nil
+}
+
+// ConstructNFT appends registry and tokenID to byte slice
+func ConstructNFT(registry common.Address, tokenID []byte) ([]byte, error) {
+	var nft []byte
+	// first 20 bytes of registry
+	nft = append(nft, registry.Bytes()...)
+
+	// next 32 bytes of the tokenID
+	nft = append(nft, tokenID...)
+
+	if len(nft) != nftByteCount {
+		return nil, errors.New("byte length mismatch")
+	}
+
+	return nft, nil
+}
+
+// isNFTInRole checks if the given nft(registry + token) is part of the core Document role.
+// If found, returns the index of the nft in the role and true
+func isNFTInRole(role *coredocumentpb.Role, registry common.Address, tokenID []byte) (nftIdx int, found bool) {
+	enft, err := ConstructNFT(registry, tokenID)
+	if err != nil {
+		return nftIdx, false
+	}
+
+	for i, n := range role.Nfts {
+		if bytes.Equal(n, enft) {
+			return i, true
+		}
+	}
+
+	return nftIdx, false
+}
+
+func getStoredNFT(nfts []*coredocumentpb.NFT, registry []byte) *coredocumentpb.NFT {
+	for _, nft := range nfts {
+		if bytes.Equal(nft.RegistryId[:20], registry) {
+			return nft
+		}
+	}
+
+	return nil
+}
+
+func getReadAccessProofKeys(cd coredocumentpb.CoreDocument, registry common.Address, tokenID []byte) (pks []string, err error) {
+	var rridx int  // index of the read rules which contain the role
+	var ridx int   // index of the role
+	var nftIdx int // index of the NFT in the above role
+	var rk []byte  // role key of the above role
+
+	found := findRole(cd, func(i, j int, role *coredocumentpb.Role) bool {
+		z, found := isNFTInRole(role, registry, tokenID)
+		if found {
+			rridx = i
+			ridx = j
+			rk = role.RoleKey
+			nftIdx = z
+		}
+
+		return found
+	}, coredocumentpb.Action_ACTION_READ)
```

Comment:
^

Reply:
These are mostly internal. We dont need to create short forms and relying on protobuf generated constants is better IMO